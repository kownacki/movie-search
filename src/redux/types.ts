export const NO_MOVIES = 'NO_MOVIES';
export type NO_MOVIES = typeof NO_MOVIES;
export const REQUEST_MOVIES = 'REQUEST_MOVIES';
export type REQUEST_MOVIES = typeof REQUEST_MOVIES;
export const SUCCESS_MOVIES = 'SUCCESS_MOVIES';
export type SUCCESS_MOVIES = typeof SUCCESS_MOVIES;
export const FAILURE_MOVIES = 'FAILURE_MOVIES';
export type FAILURE_MOVIES = typeof FAILURE_MOVIES;

export type MoviesStatus = NO_MOVIES | REQUEST_MOVIES | SUCCESS_MOVIES | FAILURE_MOVIES;
