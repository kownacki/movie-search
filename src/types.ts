export interface Movie {
  imdbID: string,
  Poster: string | null, // null means no poster available
  Title: string,
  Year: string,
}
