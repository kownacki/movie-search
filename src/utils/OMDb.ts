import {Movie} from './types';
const OMDB_API_URL = 'https://www.omdbapi.com/';

// Blank posters are described by 'N/A'. Replace them with null.
const nullifyBlankPoster = (movie: Movie): Movie => ({...movie, Poster: movie.Poster === 'N/A' ? null : movie.Poster});

const getMoviesFromResponse = async (response: Response): Promise<Movie[]> => {
  const json = await response.json();
  return (json.Response === "True") ? json.Search.map(nullifyBlankPoster) : [];
};

interface SearchResult {
  movies?: Movie[],
  isSuccess: boolean,
}

export const search = async (query: string): Promise<SearchResult> => {
  // todo add more than 10
  const response = await fetch(`${OMDB_API_URL}?s=${query}&type=movie&apikey=${process.env.REACT_APP_OMDB_API_KEY}`);
  return response.status === 200
    ? {movies: await getMoviesFromResponse(response), isSuccess: true}
    : {isSuccess: false}
};
