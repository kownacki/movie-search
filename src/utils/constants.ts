export const MAX_INPUT_LENGTH = 100;
