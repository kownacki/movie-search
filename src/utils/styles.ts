// Here are kept all common style configuration and utils used across all components.

export const borderRadius = 10;

export const cardBoxShadow = '2px 4px 10px rgba(0, 0, 0, 0.2)';

export const rwdBreakpoints = {
  1: 599,
  2: 419,
};

export const widthAndMinWidth = (width: number): string => `
  width: ${width}px;
  min-width: ${width}px;
`;
