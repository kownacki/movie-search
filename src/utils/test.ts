import * as types from './types';

export const createMovieStub = (changedProps?: Partial<types.Movie>): types.Movie => ({
  imdbID: 'id',
  Poster: 'posterStub',
  Title: 'Title Stub',
  Year: 'yearStub',
  ...changedProps,
});
