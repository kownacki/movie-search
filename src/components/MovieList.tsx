import * as React from 'react';
import * as types from '../utils/types';
import MovieCard from './MovieCard';

interface MovieListProps {
  movies: types.Movie[],
}

const renderSwitch = ({movies}: MovieListProps) => {
  if (movies.length) {
    return (
      <ul>
        {movies.map((movie) =>
          <li key={movie.imdbID}>
            <MovieCard movie={movie} />
          </li>
        )}
      </ul>
    );
  } else {
    return <div>No movies found</div>;
  }
};

const MovieList = (props: MovieListProps): React.ReactElement => {
  return renderSwitch(props);
};

export default MovieList;
