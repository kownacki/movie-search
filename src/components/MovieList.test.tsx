import React from 'react';
import {shallow} from 'enzyme';
import MovieList from './MovieList';
import * as types from '../utils/types';
import * as testUtils from '../utils/test';

describe('MovieList', () => {
  it('correctly renders movies', () => {
    const moviesStub: types.Movie[] = [
      // arbitrary number of movies with different ids
      testUtils.createMovieStub({imdbID: 'id1'}),
      testUtils.createMovieStub({imdbID: 'id2'}),
      testUtils.createMovieStub({imdbID: 'id3'}),
    ];
    const movieList = shallow(<MovieList movies={moviesStub} />);
    expect(movieList).toMatchSnapshot();
  });

  it('renders correctly when no movies', () => {
    const emptyMoviesStub: types.Movie[] = [];
    const movieList = shallow(<MovieList movies={emptyMoviesStub} />);
    expect(movieList).toMatchSnapshot();
  });
});
