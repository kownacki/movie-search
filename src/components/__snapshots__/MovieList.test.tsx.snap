// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieList correctly renders movies 1`] = `
<ul>
  <li
    key="id1"
  >
    <Styled(MovieCard)
      movie={
        Object {
          "Poster": "posterStub",
          "Title": "Title Stub",
          "Year": "yearStub",
          "imdbID": "id1",
        }
      }
    />
  </li>
  <li
    key="id2"
  >
    <Styled(MovieCard)
      movie={
        Object {
          "Poster": "posterStub",
          "Title": "Title Stub",
          "Year": "yearStub",
          "imdbID": "id2",
        }
      }
    />
  </li>
  <li
    key="id3"
  >
    <Styled(MovieCard)
      movie={
        Object {
          "Poster": "posterStub",
          "Title": "Title Stub",
          "Year": "yearStub",
          "imdbID": "id3",
        }
      }
    />
  </li>
</ul>
`;

exports[`MovieList renders correctly when no movies 1`] = `
<div>
  No movies found
</div>
`;
